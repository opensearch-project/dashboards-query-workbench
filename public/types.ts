/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { DataSourcePluginStart } from '../../../src/plugins/data_source/public/types';
import { NavigationPublicPluginStart } from '../../../src/plugins/navigation/public';
import { ObservabilityStart } from './types';

export interface WorkbenchPluginSetup {}

// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface WorkbenchPluginStart {
}

export interface AppPluginStartDependencies {
  navigation: NavigationPublicPluginStart;
  dataSource: DataSourcePluginStart;
  observabilityDashboards?: ObservabilityStart;
}

export type { ObservabilityStart } from '../../dashboards-observability/public';
