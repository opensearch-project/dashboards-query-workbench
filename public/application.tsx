/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */


import React from 'react';
import ReactDOM from 'react-dom';
import { AppMountParameters, CoreStart } from '../../../src/core/public';
import { DataSourceManagementPluginSetup } from '../../../src/plugins/data_source_management/public';
import { WorkbenchApp } from './components/app';
import { AppPluginStartDependencies } from './types';

export const renderApp = (
  { notifications, http, chrome, savedObjects }: CoreStart,
  { navigation, dataSource }: AppPluginStartDependencies,
  { appBasePath, element, setHeaderActionMenu }: AppMountParameters,
  dataSourceManagement: DataSourceManagementPluginSetup
) => {
  ReactDOM.render(
    <WorkbenchApp
      basename={appBasePath}
      notifications={notifications}
      http={http}
      navigation={navigation}
      chrome={chrome}
      savedObjects={savedObjects}
      dataSourceEnabled={!!dataSource}
      dataSourceManagement={dataSourceManagement}
      setActionMenu={setHeaderActionMenu}
    />,
    element
  );

  return () => ReactDOM.unmountComponentAtNode(element);
};
