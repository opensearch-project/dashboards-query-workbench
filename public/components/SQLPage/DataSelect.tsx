/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { EuiComboBox, EuiComboBoxOptionOption } from '@elastic/eui';
import React, { useEffect, useState } from 'react';
import { CoreStart } from '../../../../../src/core/public';

interface CustomView {
  http: CoreStart['http'];
  onSelect: (selectedItems: []) => void;
  urlDataSource: string;
  asyncLoading: boolean;
}

export const DataSelect = ({ http, onSelect, urlDataSource, asyncLoading }: CustomView) => {
  const [selectedOptions, setSelectedOptions] = useState<EuiComboBoxOptionOption[]>([
    { label: 'OpenSearch' },
  ]);
  const [options, setOptions] = useState<any[]>([]);

  const datasources = () => {
    let dataOptions: EuiComboBoxOptionOption[] = [];
    let urlSourceFound = false;
    http
      .get(`/api/get_datasources`)
      .then((res) => {
        const data = res.data.resp;

        const connectorGroups = {};

        data.forEach((item) => {
          const connector = item.connector;
          const name = item.name;

          if (connector === 'S3GLUE') {
            if (!connectorGroups[connector]) {
              connectorGroups[connector] = [];
            }

            connectorGroups[connector].push(name);
            if (name === urlDataSource) {
              urlSourceFound = true;
            }
          }
        });
        dataOptions.push({ label: 'OpenSearch' });

        for (const connector in connectorGroups) {
          if (connectorGroups.hasOwnProperty(connector)) {
            const connectorNames = connectorGroups[connector];

            dataOptions.push({
              label: connector,
              options: connectorNames.map((name) => ({ label: name })),
            });
          }
        }

        setOptions(dataOptions);
        if (urlSourceFound) {
          setSelectedOptions([{ label: urlDataSource }]);
          onSelect([{ label: urlDataSource }]);
        }
      })
      .catch((err) => {
        console.error(err);
      });
  };

  useEffect(() => {
    datasources();
  }, []);

  const handleSelectionChange = (selectedItems: any[]) => {
    if (selectedItems.length > 0) {
      setSelectedOptions(selectedItems);
      onSelect(selectedItems);
    }
  };

  return (
    <EuiComboBox
      singleSelection={{ asPlainText: true }}
      isClearable={false}
      options={options}
      selectedOptions={selectedOptions}
      onChange={(selectedItems) => handleSelectionChange(selectedItems)}
      isDisabled={asyncLoading}
    />
  );
};
