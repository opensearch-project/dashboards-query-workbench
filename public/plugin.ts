/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { AppMountParameters, CoreSetup, CoreStart, Plugin } from '../../../src/core/public';
import { DataSourcePluginSetup, DataSourcePluginStart } from '../../../src/plugins/data_source/public';
import { DataSourceManagementPluginSetup } from '../../../src/plugins/data_source_management/public';
import { PLUGIN_NAME } from '../common/constants';
import { registerObservabilityDependencies } from './dependencies/register_observability_dependencies';
import { coreRefs } from './framework/core_refs';
import { AppPluginStartDependencies, WorkbenchPluginSetup, WorkbenchPluginStart } from './types';
export interface WorkbenchPluginSetupDependencies {
  dataSource: DataSourcePluginSetup;
  dataSourceManagement: DataSourceManagementPluginSetup
}

export interface WorkbenchPluginStartDependencies {
  dataSource: DataSourcePluginStart;
}
export class WorkbenchPlugin implements Plugin<WorkbenchPluginSetup, WorkbenchPluginStart> {
  public setup(core: CoreSetup, {dataSource, dataSourceManagement} : WorkbenchPluginSetupDependencies): WorkbenchPluginSetup {
    // Register an application into the side navigation menu
    core.application.register({
      id: 'opensearch-query-workbench',
      title: PLUGIN_NAME,
      category: {
        id: 'opensearch',
        label: 'OpenSearch Plugins',
        order: 2000,
      },
      order: 1000,
      async mount(params: AppMountParameters) {
        // Load application bundle
        const { renderApp } = await import('./application');
        // Get start services as specified in opensearch_dashboards.json
        const [coreStart, depsStart] = await core.getStartServices();
        // Render the application
        return renderApp(coreStart, depsStart as AppPluginStartDependencies, params, dataSourceManagement);
      },
    });

    // Return methods that should be available to other plugins
    return {};
  }

  public start(core: CoreStart, startDeps: AppPluginStartDependencies): WorkbenchPluginStart {
    coreRefs.http = core.http;
    coreRefs.savedObjectsClient = core.savedObjects.client;
    coreRefs.toasts = core.notifications.toasts;
    coreRefs.chrome = core.chrome;
    coreRefs.application = core.application;
    coreRefs.overlays = core.overlays;
    coreRefs.dataSource = startDeps.dataSource;

    registerObservabilityDependencies(startDeps.observabilityDashboards);
    return {};
  }

  public stop() {}
}
