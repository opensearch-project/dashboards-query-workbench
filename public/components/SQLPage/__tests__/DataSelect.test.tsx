/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */


import '@testing-library/jest-dom';
import '@testing-library/jest-dom/extend-expect';
import { waitFor } from '@testing-library/react';
import { configure, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';
import React from 'react';
import { HttpResponse } from '../../../../../../src/core/public';
import { httpClientMock } from '../../../../test/mocks';
import { mockDataSelectQuery } from '../../../../test/mocks/mockData';
import { DataSelect } from '../DataSelect';

describe('Renders the Datasource picker component', () => {
  configure({ adapter: new Adapter() });

  it('fetches the datasources and selects S3glue as source', async () => {
    const client = httpClientMock;
    client.get = jest.fn(() => {
      return (Promise.resolve(mockDataSelectQuery) as unknown) as HttpResponse;
    });

    const mockOnSelect = jest.fn();
    const wrapper = mount(
      <DataSelect
        http={client}
        onSelect={mockOnSelect}
        urlDataSource={'glue_1'}
        asyncLoading={false}
      />
    );

    wrapper.update();
    await waitFor(() => {
      expect(
        toJson(wrapper, {
          noKey: false,
          mode: 'deep',
        })
      ).toMatchSnapshot();
      //   expect(screen.getByText('glue_1')).toBeInTheDocument();
    });
  });
  it('fetches the datasources and selects Opensearch as source', async () => {
    const client = httpClientMock;
    client.get = jest.fn(() => {
      return (Promise.resolve(mockDataSelectQuery) as unknown) as HttpResponse;
    });

    const mockOnSelect = jest.fn();
    const wrapper = mount(
      <DataSelect
        http={client}
        onSelect={mockOnSelect}
        urlDataSource={'OpenSearch'}
        asyncLoading={false}
      />
    );

    wrapper.update();
    await waitFor(() => {
      expect(
        toJson(wrapper, {
          noKey: false,
          mode: 'deep',
        })
      ).toMatchSnapshot();
    });
  });
});