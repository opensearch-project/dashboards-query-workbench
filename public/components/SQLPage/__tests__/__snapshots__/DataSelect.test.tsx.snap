// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders the Datasource picker component fetches the datasources and selects Opensearch as source 1`] = `
<div
  aria-expanded={false}
  aria-haspopup="listbox"
  className="euiComboBox"
  onKeyDown={[Function]}
  role="combobox"
>
  <div
    className="euiFormControlLayout"
  >
    <div
      className="euiFormControlLayout__childrenWrapper"
    >
      <div
        className="euiComboBox__inputWrap euiComboBox__inputWrap--noWrap"
        data-test-subj="comboBoxInput"
        onClick={[Function]}
        tabIndex={-1}
      >
        <span
          className="euiComboBoxPill euiComboBoxPill--plainText"
        >
          OpenSearch
        </span>
        <div
          className="euiComboBox__input"
          style={
            Object {
              "display": "inline-block",
              "fontSize": 14,
            }
          }
        >
          <input
            aria-controls=""
            data-test-subj="comboBoxSearchInput"
            disabled={false}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            role="textbox"
            style={
              Object {
                "boxSizing": "content-box",
                "width": "2px",
              }
            }
            value=""
          />
          <div
            style={
              Object {
                "height": 0,
                "left": 0,
                "overflow": "scroll",
                "position": "absolute",
                "top": 0,
                "visibility": "hidden",
                "whiteSpace": "pre",
              }
            }
          />
        </div>
      </div>
      <div
        className="euiFormControlLayoutIcons euiFormControlLayoutIcons--right"
      >
        <button
          aria-label="Open list of options"
          className="euiFormControlLayoutCustomIcon euiFormControlLayoutCustomIcon--clickable"
          data-test-subj="comboBoxToggleListButton"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          <svg
            aria-hidden={true}
            className="euiIcon euiIcon--medium euiFormControlLayoutCustomIcon__icon"
            focusable="false"
            height={16}
            role="img"
            style={null}
            viewBox="0 0 16 16"
            width={16}
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M13.069 5.157 8.384 9.768a.546.546 0 0 1-.768 0L2.93 5.158a.552.552 0 0 0-.771 0 .53.53 0 0 0 0 .759l4.684 4.61c.641.631 1.672.63 2.312 0l4.684-4.61a.53.53 0 0 0 0-.76.552.552 0 0 0-.771 0Z"
              fillRule="non-zero"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`Renders the Datasource picker component fetches the datasources and selects S3glue as source 1`] = `
<div
  aria-expanded={false}
  aria-haspopup="listbox"
  className="euiComboBox"
  onKeyDown={[Function]}
  role="combobox"
>
  <div
    className="euiFormControlLayout"
  >
    <div
      className="euiFormControlLayout__childrenWrapper"
    >
      <div
        className="euiComboBox__inputWrap euiComboBox__inputWrap--noWrap"
        data-test-subj="comboBoxInput"
        onClick={[Function]}
        tabIndex={-1}
      >
        <span
          className="euiComboBoxPill euiComboBoxPill--plainText"
        >
          OpenSearch
        </span>
        <div
          className="euiComboBox__input"
          style={
            Object {
              "display": "inline-block",
              "fontSize": 14,
            }
          }
        >
          <input
            aria-controls=""
            data-test-subj="comboBoxSearchInput"
            disabled={false}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            role="textbox"
            style={
              Object {
                "boxSizing": "content-box",
                "width": "2px",
              }
            }
            value=""
          />
          <div
            style={
              Object {
                "height": 0,
                "left": 0,
                "overflow": "scroll",
                "position": "absolute",
                "top": 0,
                "visibility": "hidden",
                "whiteSpace": "pre",
              }
            }
          />
        </div>
      </div>
      <div
        className="euiFormControlLayoutIcons euiFormControlLayoutIcons--right"
      >
        <button
          aria-label="Open list of options"
          className="euiFormControlLayoutCustomIcon euiFormControlLayoutCustomIcon--clickable"
          data-test-subj="comboBoxToggleListButton"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          <svg
            aria-hidden={true}
            className="euiIcon euiIcon--medium euiIcon-isLoading euiFormControlLayoutCustomIcon__icon"
            focusable="false"
            height={16}
            role="img"
            style={null}
            viewBox="0 0 16 16"
            width={16}
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M5.277 10.088c.02.014.04.03.057.047.582.55 1.134.812 1.666.812.586 0 1.84-.293 3.713-.88L9 6.212V2H7v4.212l-1.723 3.876Zm-.438.987L3.539 14h8.922l-1.32-2.969C9.096 11.677 7.733 12 7 12c-.74 0-1.463-.315-2.161-.925ZM6 2H5V1h6v1h-1v4l3.375 7.594A1 1 0 0 1 12.461 15H3.54a1 1 0 0 1-.914-1.406L6 6V2Z"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</div>
`;
